RIOT-OS/RIOT
# gnrc_netif: implementation of dynamic GNRC_NETIF_NUMOF approach

### Contribution description
This PR is an implementation of the ideas proposed in https://github.com/RIOT-OS/RIOT/issues/9903

It removes the internal allocation of `gnrc_netif_t` objects in gnrc_netif and assumes the `gnrc_netif_xxx_create` caller allocates the network interface. This means the allocation is decentralized and we move to a linked list approach for `gnrc_netif_numof` and `gnrc_netif_iter` (as described in #9903)

Some benefits:
- No need to do tricks for incrementing the number of interfaces. Just call `gnrc_netif_xxx_create` and they will be in the linked list :)
- Auto init can define the number of interfaces depending of the number of devices (e.g 2 IEEE802.15.4 interfaces if there are 2 AT86RF2XX radios)
- A little bit less of ROM and RAM, since it's safe to assume the pointers are always there.

I added a `GNRC_NETIF_SINGLE` macro (as proposed in #9903) to keep the optimizations when there's a single interface.

### Testing procedure
This should be carefully tested. Probably 95% of testing here is just compile tests, but it would be nice to ensure that the single interface optimization (`GNRC_NETIF_SINGLE`) still works as the original `GNRC_NETIF_NUMOF==1`.

Also, ensure that the devices still work.



### Issues/PRs references
#9903 
Alternative to https://github.com/RIOT-OS/RIOT/pull/12308
Fixes stuff like https://github.com/RIOT-OS/RIOT/pull/11979
